### PR TITLE
WAM Fix for ADFS 2019

### LINKED ISSUE #2836: [Bug] ADFS 2019 + WAM throws exception
**Which version of MSAL.NET are you using?**
MSAL.Net v4.35.1

**Platform**
.Net WPF app

**What authentication flow has the issue?**
* Desktop / Mobile
    * [ ] Interactive
    * [ x] Integrated Windows Authentication
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web app 
    * [ ] Authorization code
    * [ ] On-Behalf-Of
* Daemon app 
    * [ ] Service to Service calls

**Is this a new or existing app?**
new

**Repro**
This is pure ADFS 2019 environment (no Azure AD involved).  I am trying to use MSAL with WAM  to do Integrated Windows Authentication to ADFS as suggested in https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2771#issuecomment-890963622 but this does not work.

 I am getting the following exception calling AcquireTokenInteractive using WAM:

System.ArgumentNullException
  HResult=0x80004003
  Message=Value cannot be null.
  Source=mscorlib
  StackTrace:
   at System.StubHelpers.HStringMarshaler.ConvertToNativeReference(String managed, HSTRING_HEADER* hstringHeader)
   at Windows.Security.Authentication.Web.Core.WebAuthenticationCoreManager.FindAccountProviderAsync(String webAccountProviderId, String authority)
   at Microsoft.Identity.Client.Platforms.Features.WamBroker.WebAccountProviderFactory.<GetAccountProviderAsync>d__0.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.Platforms.Features.WamBroker.WamBroker.<GetProviderAsync>d__23.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.Identity.Client.Platforms.Features.WamBroker.WamBroker.<AcquireTokenInteractiveAsync>d__13.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.Internal.Broker.BrokerInteractiveRequestComponent.<FetchTokensAsync>d__9.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.<FetchTokensFromBrokerAsync>d__10.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.<GetTokenResponseAsync>d__11.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.Internal.Requests.InteractiveRequest.<ExecuteAsync>d__8.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.<RunAsync>d__12.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.Identity.Client.ApiConfig.Executors.PublicClientExecutor.<ExecuteAsync>d__2.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at active_directory_wpf_msgraph_v2.MainWindow.<CallGraphButton_Click>d__3.MoveNext() in C:\Users\bhadmin.AZUREAD\Documents\sample\WAM\active-directory-dotnet-desktop-msgraph-v2\active-directory-wpf-msgraph-v2\MainWindow.xaml.cs:line 78

 
  This exception was originally thrown at this call stack:
    [External Code]
    active_directory_wpf_msgraph_v2.MainWindow.CallGraphButton_Click(object, System.Windows.RoutedEventArgs) in MainWindow.xaml.cs
 
 
My code (modified from sample:  https://github.com/Azure-Samples/active-directory-dotnet-desktop-msgraph-v2:
 
 registered redirect URI in ADFS 3019 is  "ms-appx-web://microsoft.aad.brokerplugin/{client_id}" for Native Desktop app

            var builder = PublicClientApplicationBuilder.Create("<client ID>")
                .WithAdfsAuthority("https://<adfs domain>/adfs/", false)
                .WithDefaultRedirectUri();
                builder.WithWindowsBroker(true);  // Requires redirect URI "ms-appx-web://microsoft.aad.brokerplugin/{client_id}" in app registration

            _clientApp = builder.Build();
            TokenCacheHelper.EnableSerialization(_clientApp.UserTokenCache);

 
...

                    firstAccount = PublicClientApplication.OperatingSystemAccount;

 try
            {
                authResult = await app.AcquireTokenSilent(scopes, firstAccount)
                    .ExecuteAsync();
            }
            catch (MsalUiRequiredException ex)
            {
                // A MsalUiRequiredException happened on AcquireTokenSilent. 
                // This indicates you need to call AcquireTokenInteractive to acquire a token
                System.Diagnostics.Debug.WriteLine($"MsalUiRequiredException: {ex.Message}");

 

                try
                {
                    authResult = await app.AcquireTokenInteractive(scopes)
                        .WithAccount(firstAccount)
                        .WithParentActivityOrWindow(new WindowInteropHelper(this).Handle) // optional, used to center the browser on the window
                        .WithPrompt(Prompt.SelectAccount)
                        .ExecuteAsync();
                }
                catch (MsalException msalex)
                {
                    ResultText.Text = $"Error Acquiring Token:{System.Environment.NewLine}{msalex}";
                }

**Expected behavior**
no exception and IWA authentication should work

**Actual behavior**
exception above

**Possible solution**
none for IWA

**Additional context / logs / screenshots**
IWA in my environment works fine outside of MSAL.  I can use the browser to do IWA to ADFS server correctly.

### PR DESCRIPTION
**Pull request by gladjohn**
Fixes #2836 

**Changes proposed in this request**
WAM is not supported in pure ADFS environments. We do not get an ID token from WAM for ADFS scenarios, hence we are not able to properly form the TokenResponse object and also caching fails. TO fix the issue, we are falling back to browser if the Authority is set as ADFS. 

**Testing**
Unit Tests, Manual Testing. For Release 4.41.0 this needs a manual testing on an ADFS 2019 domain joined machine

**Performance impact**
N/A

### CODE REVIEW

**Review by SameerK-MSFT — APPROVED** [review]
Please make sure that this does not break any existing B2C apps.

File: src/client/Microsoft.Identity.Client/Platforms/iOS/Broker/iOSBroker.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ public iOSBroker(ICoreLogger logger, ICryptographyManager cryptoManager, CoreUIP
             _uIParent = uIParent;
         }
 
-        public bool IsBrokerInstalledAndInvokable()
+        public bool IsBrokerInstalledAndInvokable(AuthorityType authorityType = AuthorityType.Aad)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Platforms/iOS/Broker/iOSBroker.cs:50]: Discussed offline - would be clearer if this param didn't have a default value and so callers will pass the actual type. Android and iOS brokers would just ignore the type in their implementation of `IsBrokerInstalledAndInvokable` (i.e. only the signature changes, but the logic stays the same), to minimize the changes .

File: tests/Microsoft.Identity.Test.Unit/AppConfigTests/PublicClientApplicationBuilderTests.cs [hunk 2]
```diff
@@ -640,19 +641,50 @@ public void IsBrokerAvailable_net5()
...
+                   .WithDesktopFeatures()
+                   .WithAuthority(TestConstants.AuthorityTenant);
 
 
             // broker is not available out of the box
             Assert.AreEqual(DesktopOsHelper.IsWin10OrServerEquivalent(), builder2.IsBrokerAvailable());
         }
+
+        [TestMethod]
+        public void NoBrokerADFS_OldDotNet()
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Unit/AppConfigTests/PublicClientApplicationBuilderTests.cs:661]: This test will run on: .net classic, .net core and .net5 (which is probably what you want).

**Review by neha-bhargava — APPROVED** [review]
👍

File: src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs [hunk 3]
```diff
@@ -698,14 +698,21 @@ public void HandleInstallUrl(string appLink)
...
 #if NET_CORE
             if (!DesktopOsHelper.IsWindows())
             {
                 return false;
             }
 #endif            
+            // WAM does not work on pure ADFS environments
+            if (authorityType == AuthorityType.Adfs)
+            {
+                _logger.Info("[WAM Broker] Falling Back to Browser for ADFS Authority. ");
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs:712]: added

File: src/client/Microsoft.Identity.Client/Platforms/iOS/Broker/iOSBroker.cs (hunk 1 above)
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/Platforms/iOS/Broker/iOSBroker.cs:50]: Ok, made changed

**Review by pmaytak — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs [hunk 4]
```diff
@@ -698,14 +698,24 @@ public void HandleInstallUrl(string appLink)
             throw new NotImplementedException();
         }
 
-        public bool IsBrokerInstalledAndInvokable()
+        public bool IsBrokerInstalledAndInvokable(AuthorityType authorityType = AuthorityType.Aad)
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs:701]: it does not have to removed it

File: src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs [hunk 5]
```diff
@@ -698,14 +698,24 @@ public void HandleInstallUrl(string appLink)
...
         {
 #if NET_CORE
             if (!DesktopOsHelper.IsWindows())
             {
+                _logger.Info("[WAM Broker] Not a Windows Operating System. WAM Broker will not be used. ");
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/WamBroker.cs:706]: modified

File: src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerFactory.cs [hunk 6]
```diff
@@ -33,7 +33,9 @@ private static async Task<IBroker> GetInstalledBrokerAsync(CoreUIParent uIParent
         {
             AndroidBrokerHelper brokerHelper = new AndroidBrokerHelper(Application.Context, logger);
 
-            if (brokerHelper.IsBrokerInstalledAndInvokable())
+            AuthorityType authorityType = AuthorityType.Aad; //passing AAD as default authority
+
+            if (brokerHelper.IsBrokerInstalledAndInvokable(authorityType))
```
[the remark this answers is not in the document]
Comment (gladjohn) [reply src/client/Microsoft.Identity.Client/Platforms/Android/Broker/AndroidBrokerFactory.cs:38]: added

**Pull request merged**